scripts/core/PROG.rst

```diff
@@ -1714,6 +1723,39 @@ The following table documents the supported knobs for overriding default functio
 | Memory          | ${X}_SHARED_FORCE_DEVICE_ALLOC        | {**0**, 1} | Forces all shared allocations into device memory                                  |
 +-----------------+-------------------------------------+------------+-----------------------------------------------------------------------------------+
 
+%endif
+
+%if ver >= 1.7:
++-----------------+-------------------------------------+-----------------+-----------------------------------------------------------------------------------+
+| Category        | Name                                | Values          | Description                                                                       |
++=================+=====================================+=================+===================================================================================+
+| Device          | ${X}_FLAT_DEVICE_HIERARCHY          | {**0**, 1, 2}   | Defines device hierarchy model exposed by Level Zero driver implementation        |
++                 +-------------------------------------+-----------------+-----------------------------------------------------------------------------------+
+|                 | ${X}_AFFINITY_MASK                  | list            | Forces driver to only report devices (and sub-devices) as specified by values     |
++                 +-------------------------------------+-----------------+-----------------------------------------------------------------------------------+
+|                 | ${X}_ENABLE_PCI_ID_DEVICE_ORDER     | {**0**, 1}      | Forces driver to report devices from lowest to highest PCI bus ID                 |
++-----------------+-------------------------------------+-----------------+-----------------------------------------------------------------------------------+
+| Memory          | ${X}_SHARED_FORCE_DEVICE_ALLOC      | {**0**, 1}      | Forces all shared allocations into device memory                                  |
++-----------------+-------------------------------------+-----------------+-----------------------------------------------------------------------------------+
+%endif
+
+%if ver >= 1.7:
+
+Device Hierarchy
+~~~~~~~~~~~~~
+
+${X}_FLAT_DEVICE_HIERARCHY allows users to select the device hierarchy model with which the underlying hardware is exposed and the types of devices returned with ${x}DeviceGet.
+
+With a value of `0`, ${x}DeviceGet returns all the devices that do not have a root-device. Traversing the device hierarchy is possible by querying sub-devices with ${x}DeviceGetSubDevices and root-devices with ${x}DeviceGetRootDevice. Driver implementation may perform implicit optimizations to submissions and allocations done in the root-devices.
```

Comment:
in mode 0 DeviceGetRootDevice is not allowed.

Reply:
@MichalMrozek 

You mean in mode 0, which is what we have currently, the application cannot query for the root device? could you elaborate on why would that be? this has been a request from some customers some times, to query for the parent device for a sub-device. From point of view of implementation, this doesnt change too much, as the Level Zero Driver implementation already has that info, it would be just a matter of returning it.

Reply:
The API was only meant for mode 2, when you start to flatten the hierarchy.
If hierarchy is not flattened (as in mode 0), then getroot is not supported.

That's the whole difference between mode 2 and other modes, it allows this new API.

Reply:
@MichalMrozek : thanks. yes, the API was originally meant for mode 2, but there's no functional or performance reason to not have it available in mode 0, would that be correct? By having it in mode 0 as well, we allow users to have more flexibility and broader set of APIs to use.

Reply:
let's say you are in mode 0 
app did ZE_AFFINITY_MASK=0.0 
app calls getRoot on such device, what is returned?

Reply:
@MichalMrozek it is return nullptr, but if you are in mode 0, and use ZE_AFFINITY_MASK=0, then you can call getRootDevice on any handle obtained with getSubDevices and the the parent device. It is the same for getSubDevices. With MASK=0, getSubDevices returns a list, and with MASK=0.0 getSubDevices return a count of 0. Spec doesnt forbid the use of getSubDevices with MASK=0.0, it just returns the appropriate value.

Reply:
but in mode 1 , we will return all sub devices as root device and getRoot will return nullptr?